[MLIR] Add lowering for TF::SeluOp and TF::SeluGradOp

pr4tgpt opened the pull request:
This PR adds the lowering for TF::SeluOp and TF::SeluGradOp to mhlo. Relevant test cases are also added.

This PR also adds TF to TF lowering for the same.

Signed-Off-By: Prateek Gupta <prateek@polymagelabs.com>

pr4tgpt commented:
Hi @smit-hinsu can you please review this? Thanks!

pr4tgpt commented:
> Hi @smit-hinsu can you please review this? Thanks!

Gentle ping, @smit-hinsu .

pr4tgpt commented:
Gentle ping, @joker-eph @sanjoy.

smit-hinsu commented on tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc:
```diff
@@ -1446,6 +1446,120 @@ class ConvertRollOp : public OpRewritePattern<TF::RollOp> {
+        loc, featureType, oneValue, featureShapeAttr);
+
+    // Get the value of exp(x).
+    Value expValue = rewriter.create<ExpOp>(loc, featureType, features);
```
use exp_value to use the google3 style.

smit-hinsu commented on tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc:
```diff
@@ -1446,6 +1446,120 @@ class ConvertRollOp : public OpRewritePattern<TF::RollOp> {
+/// scaledAlpha(`kScaledAlpha`) is taken as a
+/// constant 1.7580993408473768599402175208123. This choice of constants is in
+/// accordance with existing TF2XLA bridge.
+class ConvertSeluOp : public OpRewritePattern<TF::SeluOp> {
```
We should move this to lower_tf.cc instead of directly lowering to mhlo. That way the pattern is useful in other contexts as well.

I think it should have been possible to implement these patterns using tablegen. Did you consider that?

If you are in a hurry, we can add TODO for these now.

pr4tgpt replied:
Hi @smit-hinsu I have added these as DRRs. Please re-review.

smit-hinsu replied:
Did you consider moving the pattern to lower_tf so that other backends other than mhlo can make use of the pattern?
https://github.com/tensorflow/tensorflow/blob/c83cdf85c5937730af51a384c0cc605bfb8f9214/tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.cc

This could also simplify the pattern by making use of other TensorFlow ops.

This shouldn't be difficult but if you need to commit this sooner for some deadline, we can add a TODO and I can review this.

pr4tgpt replied:
@smit-hinsu I have started work on this. Will update the PR soon. Thanks!

smit-hinsu reviewed (commented):
Let's remove these ops from this list so that this patterns are used in all contexts:
https://github.com/tensorflow/tensorflow/blob/c83cdf85c5937730af51a384c0cc605bfb8f9214/tensorflow/compiler/mlir/xla/transforms/legalize_tf_with_tf2xla.cc#L226

I am also curious about your usecase that the above didn't work.

Sorry for the delay on this. Thanks for the contributions.

pr4tgpt commented:
> Let's remove these ops from this list so that this patterns are used in all contexts:
> https://github.com/tensorflow/tensorflow/blob/c83cdf85c5937730af51a384c0cc605bfb8f9214/tensorflow/compiler/mlir/xla/transforms/legalize_tf_with_tf2xla.cc#L226
> 
> I am also curious about your usecase that the above didn't work.
> 
> Sorry for the delay on this. Thanks for the contributions.

Please re-review @smit-hinsu. I have used DRRs now for the lowering. I am not sure about the code you have pointed. Can you elaborate on this?

smit-hinsu commented on tensorflow/compiler/mlir/xla/transforms/legalize_tf_patterns.td:
```diff
@@ -491,6 +491,48 @@ def : Pat<(TF_EluGradOp AnyStaticShapeTensor:$gradients, AnyRankedTensor:$featur
+           (HLOClient_BroadcastMulOp
+            $features,
+            (HLO_ConstOp: $scale
+              (NativeCodeCall<"$_builder.getFloatAttr(((*features.begin()).getType()).cast<mlir::RankedTensorType>().getElementType(), 1.0507009873554804934193349852946)">)),
```
we can define this outside to keep the pattern concise? See Get2DTransposePerm in this file. 

Also, use getElementTypeOrSelf helper.

pr4tgpt replied:
Done.

smit-hinsu commented:
> > Let's remove these ops from this list so that this patterns are used in all contexts:
> > https://github.com/tensorflow/tensorflow/blob/c83cdf85c5937730af51a384c0cc605bfb8f9214/tensorflow/compiler/mlir/xla/transforms/legalize_tf_with_tf2xla.cc#L226
> > 
> > I am also curious about your usecase that the above didn't work.
> > Sorry for the delay on this. Thanks for the contributions.
> 
> Please re-review @smit-hinsu. I have used DRRs now for the lowering. I am not sure about the code you have pointed. Can you elaborate on this?

How do you use the legalize pass? Do you have xla-legalize-tf-with-tf2xla pass in the pass pipeline?

xla-legalize-tf-with-tf2xla pass in that file is used to legalize ops mentioned in the list. That pass can already legalize Selu op but now that we will have native lowering after this change, we can drop it from the list to keep only one implementation. So, just remove Selu op from the list.

gbaned commented:
@pr4tgpt  Any update on this PR? Please. Thanks!

pr4tgpt commented:
@gbaned I am currently working on the changes mentioned by @smit-hinsu in the ``lower_tf.cc``. Will soon this update this PR.

pr4tgpt committed:
Add TF to TF lowering of TF::SeluOp and TF::SeluGradOp.

This commit adds the TF to TF lowering for Tf::SeluOp and TF::SeluGradOp.
This lowering breaks the complex TF function into simpler TF function.
Relevant test cases are also added.

Signed-Off-By: Prateek Gupta <prateek@polymagelabs.com>

pr4tgpt commented:
Hi @smit-hinsu please re-review. I have added changes as you suggested. Thanks!

pr4tgpt commented:
Gentle ping @smit-hinsu.

smit-hinsu commented:
There were holidays in the US so will get to this on Tuesday or Wednesday.

pr4tgpt commented:
> There were holidays in the US so will get to this on Tuesday or Wednesday.

Gentle Ping @smit-hinsu .

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_generated_ops.td:
```diff
@@ -4434,7 +4434,7 @@ dimension of size 1 added.}]>:$output
 }
 
-def TF_Expm1Op : TF_Op<"Expm1", [NoSideEffect, TF_SameOperandsAndResultTypeResolveRef]> {
+def TF_Expm1Op : TF_Op<"Expm1", [NoSideEffect, SameOperandsAndResultType]> {
```
why is this change from TF_SameOperandsAndResultTypeResolveRef to SameOperandsAndResultType?

pr4tgpt replied:
``TF_SameOperandsAndResultTypeResolveRef`` is somehow not accepting the pattern ``TF_Expm1Op $features``. It is failing the build for this.

smit-hinsu replied:
Ah, I see.

Rewrite patterns have better support for SameOperandsAndResultType trait for ops that don't have shape inference function. Declaring shape inference function for this op should make it work.

smit-hinsu replied:
Seems like this is still outdated?

pr4tgpt replied:
yes I am working on this. Will soon update this.

pr4tgpt replied:
@smit-hinsu can you elaborate on shape inference function? Are you referring to ``InferShapedTypeOpInterface``?

pr4tgpt replied:
@smit-hinsu Added the shape inference function. Please re-review. Thanks!

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -457,3 +457,45 @@ def LowerIsFiniteOp : Pat<(TF_IsFiniteOp $x),
+              $features,
+              (TF_FillOp
+                (CreateTFShapeOp $src_op, $features, /*use 32bit*/ConstBoolAttrFalse),
+                (TF_ConstOp (GetScalarOfFloatType<"1.7580993408473768599402175208123"> $features))))))>;
```
add a new line at the end of file.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -457,3 +457,45 @@ def LowerIsFiniteOp : Pat<(TF_IsFiniteOp $x),
+// Selu op patterns.
+//===----------------------------------------------------------------------===//
+
+def : Pat<(TF_SeluOp: $src_op AnyRankedTensor:$features),
```
Add a comment with the formula similar to the other ops.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -457,3 +457,45 @@ def LowerIsFiniteOp : Pat<(TF_IsFiniteOp $x),
+
+def : Pat<(TF_SeluOp: $src_op AnyRankedTensor:$features),
+        (TF_SelectOp
+          (TF_GreaterEqualOp
```
this should be greater?

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -457,3 +457,45 @@ def LowerIsFiniteOp : Pat<(TF_IsFiniteOp $x),
+        (TF_SelectOp
+          (TF_GreaterEqualOp
+            $features,
+            (TF_FillOp
```
greater supports broadcasting so fill shouldn't be necessary.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -457,3 +457,45 @@ def LowerIsFiniteOp : Pat<(TF_IsFiniteOp $x),
+              (TF_ConstOp (GetScalarOfType<0> $features)))),
+          (TF_MulOp
+            $features,
+            (TF_FillOp
```
same here

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -457,3 +457,45 @@ def LowerIsFiniteOp : Pat<(TF_IsFiniteOp $x),
+            $gradients,
+            (TF_FillOp
+              (CreateTFShapeOp $src_op, $features, /*use 32bit*/ConstBoolAttrFalse),
+              (TF_ConstOp (GetScalarOfFloatType<"1.0507009873554804934193349852946"> $features)))),
```
define a const helper for these so that the same constant doesn't need to be defined twice. We might be able to reuse GetScalarOfFloatType for this.

pr4tgpt replied:
Done. Please re-review @smit-hinsu

smit-hinsu commented on tensorflow/compiler/mlir/xla/transforms/legalize_tf_patterns.td:
```diff
@@ -491,6 +491,49 @@ def : Pat<(TF_EluGradOp AnyStaticShapeTensor:$gradients, AnyRankedTensor:$featur
                (HLO_ConstOp:$one (GetScalarOfType<1> $features)),
                (BinBroadcastDimensions $one, $features))))>;
 
+//===----------------------------------------------------------------------===//
```
remove these patterns now.

smit-hinsu commented on tensorflow/compiler/mlir/xla/transforms/legalize_tf_with_tf2xla.cc:
```diff
@@ -220,8 +220,6 @@ bool IsOpAllowedTf2XlaFallback(Operation* op) {
     TypeID::get<TF::RoundOp>(),
     TypeID::get<TF::SelectV2Op>(),
     TypeID::get<TF::SelfAdjointEigV2Op>(),
```
add these two ops in the following list IsOpAllowedTf2XlaPreferred.

smit-hinsu reviewed (commented):
We are almost there.

pr4tgpt committed:
[MLIR] Add shape inference for TF::Expm1Op.

This commit adds shape inference function for TF::Expm1Op.

Signed-Off-By: Prateek Gupta <prateek@polymagelabs.com>

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -2215,6 +2215,19 @@ void ExpandDimsOp::build(OpBuilder &builder, OperationState &result,
+    MLIRContext* context, Optional<Location> location, ValueRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<Type>& inferredReturnTypes) {
+  const TensorType input_ty = operands[0].getType().template cast<TensorType>();
```
nit: don't assume that input type is a tensor type. We generally assume in other places as the IR is verified but here we should fail if the input is not a tensor type.

smit-hinsu replied:
check that there are at least one operands.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -2215,6 +2215,19 @@ void ExpandDimsOp::build(OpBuilder &builder, OperationState &result,
+    SmallVectorImpl<Type>& inferredReturnTypes) {
+  const TensorType input_ty = operands[0].getType().template cast<TensorType>();
+  inferredReturnTypes.assign(
+      {RankedTensorType::get(input_ty.getShape(), input_ty.getElementType())});
```
getShape fails for unranked inputs.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -2215,6 +2215,19 @@ void ExpandDimsOp::build(OpBuilder &builder, OperationState &result,
+  const TensorType input_ty = operands[0].getType().template cast<TensorType>();
+  inferredReturnTypes.assign(
+      {RankedTensorType::get(input_ty.getShape(), input_ty.getElementType())});
+  return success();
```
why can't we simply forward operands[0].getType() after checking that there is at least one operand?

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -457,3 +457,49 @@ def LowerIsFiniteOp : Pat<(TF_IsFiniteOp $x),
+// constant 1.0507009873554804934193349852946 and scaledAlpha is taken
+// as a constant 1.7580993408473768599402175208123.
+
+def LowerSeluGradOp : Pat<(TF_SeluGradOp:$src_op AnyStaticShapeTensor:$gradients, TF_FloatTensor:$features),
```
make this fit in 80 chars like other patterns.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -457,3 +457,49 @@ def LowerIsFiniteOp : Pat<(TF_IsFiniteOp $x),
+
+def LowerSeluOp : Pat<(TF_SeluOp: $src_op TF_FloatTensor:$features),
+                  (TF_SelectOp
+                    (TF_GreaterEqualOp
```
this should be greater instead of GreaterEqual

https://github.com/tensorflow/tensorflow/blob/818120e208d2b5622e7988f902a58ab5df3b69d0/tensorflow/compiler/tf2xla/kernels/elu_op.cc#L70

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -457,3 +457,49 @@ def LowerIsFiniteOp : Pat<(TF_IsFiniteOp $x),
+// Selu op patterns.
+//===----------------------------------------------------------------------===//
+
+// Selu(x) = scaledAlpha * (exp(x) - 1) if x < 0 else scale * x. Here
```
update this formula to match what you have in the pattern.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -457,3 +457,49 @@ def LowerIsFiniteOp : Pat<(TF_IsFiniteOp $x),
+
+def LowerSeluGradOp : Pat<(TF_SeluGradOp:$src_op AnyStaticShapeTensor:$gradients, TF_FloatTensor:$features),
+                      (TF_SelectOp
+                        (TF_GreaterEqualOp
```
same here as well.

gbaned commented:
@pr4tgpt Can you please check @smit-hinsu's comments and keep us posted ? Thanks!

pr4tgpt commented:
> @pr4tgpt Can you please check @smit-hinsu's comments and keep us posted ? Thanks!

Sure. I am working on this. Will update the PR soon.

pr4tgpt commented:
@smit-hinsu Please re-review.

smit-hinsu reviewed (commented):
Last set of comments.

pr4tgpt commented:
Hi @smit-hinsu, Thank you for your valuable comments. Addressed all. Please re-review. Thanks!

pr4tgpt commented:
Gentle Ping @smit-hinsu .

pr4tgpt commented:
Rebased and pushed.

pr4tgpt commented:
Gentle ping @smit-hinsu.

pr4tgpt commented:
Hi @smit-hinsu,  I am not sure why tests are failing, can you tell where it is failing? Thanks!

gbaned commented:
@pr4tgpt  Can you please resolve conflicts? Thanks!

smit-hinsu commented:
Look at the failure log in this: https://tensorflow-ci.intel.com/job/tensorflow-mkl-linux-cpu-pr/10251/

pr4tgpt commented:
Hi @smit-hinsu, rebased and merged.

pr4tgpt commented:
Hi @smit-hinsu can you look into these failing tests. I dont think it is related to my code.
